pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -402,49 +361,145 @@ func (s *NodeController) checkNodeReady(node *api.Node) *api.NodeCondition {
+// a long period of time.
+func (s *NodeController) MonitorNodeStatus() error {
+	nodes, err := s.kubeClient.Nodes().List()
 	if err != nil {
 		return err
 	}
-	for _, pod := range pods.Items {
-		if pod.Status.Host != nodeID {
-			continue
+	for i := range nodes.Items {
+		var gracePeriod time.Duration
+		var lastReadyCondition api.NodeCondition
+		node := &nodes.Items[i]
+		readyCondition := s.getCondition(node, api.NodeReady)
+		if readyCondition == nil {
+			// If ready condition is nil, then kubelet (or nodecontroller) never posted node status.
+			// A fake ready condition is created, where LastProbeTime and LastTransitionTime is set
+			// to node.CreationTimestamp to avoid handle the corner case.
+			lastReadyCondition = api.NodeCondition{
+				Type:               api.NodeReady,
+				Status:             api.ConditionUnknown,
+				LastProbeTime:      node.CreationTimestamp,
+				LastTransitionTime: node.CreationTimestamp,
+			}
+			gracePeriod = nodeStartupGracePeriod
+		} else {
+			// If ready condition is not nil, make a copy of it, since we may modify it in place later.
+			lastReadyCondition = *readyCondition
+			gracePeriod = nodeMonitorGracePeriod
 		}
-		glog.V(2).Infof("Delete pod %v", pod.Name)
-		if err := s.kubeClient.Pods(pod.Namespace).Delete(pod.Name); err != nil {
-			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
+
+		// Check last time when NodeReady was updated.
+		if s.now().After(lastReadyCondition.LastProbeTime.Add(gracePeriod)) {
+			// NodeReady condition was last set longer ago than gracePeriod, so update it to Unknown
+			// (regardless of its current value) in the master, without contacting kubelet.
+			if readyCondition == nil {
+				glog.V(2).Infof("node %v is never updated by kubelet")
+				node.Status.Conditions = append(node.Status.Conditions, api.NodeCondition{
+					Type:               api.NodeReady,
+					Status:             api.ConditionUnknown,
+					Reason:             fmt.Sprintf("Kubelet never posted node status"),
+					LastProbeTime:      s.now(),
+					LastTransitionTime: s.now(),
+				})
+			} else {
+				// Note here the out-dated condition can be the one posted by nodecontroller
+				// itself before. We keep posting the status to keep LastProbeTime fresh.
+				glog.V(2).Infof("node %v hasn't been updated for a while, last ready condition is %+v", node.Name, readyCondition)
+				readyCondition.Status = api.ConditionUnknown
+				readyCondition.Reason = fmt.Sprintf("Kubelet stopped posting node status")
+				readyCondition.LastProbeTime = s.now()
+				if lastReadyCondition.Status != api.ConditionUnknown {
+					readyCondition.LastTransitionTime = s.now()
+				}
+			}
+			_, err = s.kubeClient.Nodes().Update(node)
+			if err != nil {
+				glog.Errorf("error updating node %s: %v", node.Name, err)
+			}
 		}
-	}
 
+		if readyCondition != nil {
+			// Check eviction timeout.
+			if lastReadyCondition.Status == api.ConditionNone &&
```

Comment:
This is definitely better, but I think there is still one corner case that is problematic. What if a node flips back and forth between ConditionNone and ConditionUnknown, staying in each state for < podEvictionTimeout-gracePeriod seconds. (For example, node reports not ready, then node becomes unreachable, then node becomes reachable and reports not ready, then node becomes unreachable, etc. etc.) Then I think we will never evict the pod.

I think the logic needs to be:
if ReadyCondition is ConditionNone && now > LastTransitionTime + podEvictionTimeout { deletePods }
if ReadyCondition is ConditionUnknown && now > LastProbeTime + (podEvictionTimeout-gracePeriod) { deletePods }

In other words -- if you have been in ConditionNone for podEvictionTimeout OR the last time you heard from the Kubelet was > podEvictionTimeout-gracePeriod, then evict.

What do you think?

One thing this requires is that we don't "refresh" LastProbeTime except when we actually hear from Kubelet. I didn't understand why you were "refreshing" LastProbeTime anyway. Maybe you can explain.